### PR TITLE
neptune-release Gitlab Port

### PR DESCRIPTION
**Pull request by mathias-baumann-sociomantic**
This is a preview, not intended to be merged, but open to feedback.
Things missing here are:

* extensive testing
* port of the test suite

### CODE REVIEW

File: src/octod/core.d [hunk 1]
```diff
@@ -35,6 +35,14 @@ struct Configuration
     string oauthToken;
     /// By default client works in live mode
     bool dryRun = false;
+
+    /// Returns true if this is a github configuration
+    public bool isGithub ( ) const
```
Comment (mihails-strasuns) [inline src/octod/core.d]: I know there are only two supported platforms but still it would be cleaner to have `PlatformKind platform()` method which returns enum value :)

File: src/octod/core.d [hunk 2]
```diff
@@ -334,6 +334,55 @@ struct HTTPConnection
...
+
+        Params:
+            url = GitHub API method URL (relative)
+            json = request body to send
+            accept = optional, request media type
+
+        Returns:
+            Json body of the response.
+     **/
+    Json put ( string url, Json json, MediaType accept )
```
Comment (mihails-strasuns) [inline src/octod/core.d:357]: Putting this into `octod` is a bit confusing - I think whole connection helper should be moved to own separate module and octod package removed.

File: src/octod/api/Repos.d [hunk 3]
```diff
@@ -418,6 +418,8 @@ private Repository newRepo ( ref HTTPConnection con, Json json, string repo )
 
     if (con.isGithub())
         return new GithubRepo(&con, json);
+    else
+        return new GitlabRepo(&con, json, repo);
```
Comment (mihails-strasuns) [inline src/octod/api/Repos.d]: Will be able to use `final swtich` here with my enum proposal

File: src/octod/api/repos/ReposGitlab.d [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+    /** C'tor
+
+        Params:
+            con = http connection
+            json = initial json info
+            repo = repo name
+
+     **/
+    public this ( HTTPConnection* con, const Json json, string repo )
```
Comment (mihails-strasuns) [inline src/octod/api/repos/ReposGitlab.d]: So it looks like gitlab doesn't have single "repo" json with all the data and you have to make request for individual methods, correct? In that case there is no point in keepin `Json` field in `Repository` base class - make it specifc to GitHub implementation.

File: src/internal/RemoteConfig.d [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    ***************************************************************************/
+
+    public this ( string app_name, string base_url, bool assume_yes )
+    {
+        import internal.git.helper;
+        import std.exception;
+
+        this.app_name = app_name;
+
+        immutable GitHubConfName = ".oauthtoken";
```
Comment (mihails-strasuns) [inline src/internal/RemoteConfig.d]: I think we should make this forward-compatible and also check for `.github-oauthtoken` with an assumption that git-hub will also eventually becomes generic.

File: src/release/options.d [hunk 6]
```diff
@@ -113,3 +113,49 @@ Options parseOpts ( string[] opts )
...
+*******************************************************************************/
+
+public string extractDomain ( string url )
+{
+    import std.algorithm;
+    import std.range : empty;
+
+    auto begin = url.findSplitAfter("@");
+
+    if (!begin)
```
Comment (mihails-strasuns) [inline src/release/options.d]: You don't like `enforce`, do you? :)

**Review by mihails-strasuns — COMMENTED** [review]
Looks very reasonable and pretty much what I would expect from this kind of design change. Probably biggest meta-issue is that there are plenty of leftovers of old architecture and package structure that don't make much sense anymore. After tests are in place would be nice to revisit it all and clean all the legacy.

File: src/release/options.d (hunk 6 above)
Comment (mathias-baumann-sociomantic) [reply src/release/options.d]: It's just not a pattern I am used too :-)

Comment (mathias-baumann-sociomantic) [reply src/release/options.d]: It's just not a pattern I am used too :-)

Comment (mathias-baumann-sociomantic) [reply src/release/options.d]: It's just not a pattern I am used too :-)

File: src/octod/api/common.d [hunk 7]
```diff
@@ -51,7 +51,7 @@ void validateRepoString ( string repo )
     import std.regex;
     import std.exception : enforce;
 
-    static rgxRepo = regex(r"^[^/]+/[^/]+$");
+    static rgxRepo = regex(r"^[^@/]+/[^/]+$");
```
Comment (leandro-lucarella-sociomantic) [inline src/octod/api/common.d:54]: Is this really related to this commit? Does it worth a comment explaining which is the expected form?

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
On a quick look, it LGTM.

About "Analyses the upstream URL to find out if it's github or a (possibly custom) gitlab server.
Uses the according backend then.", I support this but I think a config option should be provided where the user can override this, there is also GitHub Enterprise so with custom servers only the user could say for sure if GL or GH should be used.

File: src/octod/api/common.d (hunk 7 above)
Comment (mathias-baumann-sociomantic) [reply src/octod/api/common.d:54]: I thought it was required for it to work but I realize now that I might have been wrong

**Comment by mathias-baumann-sociomantic** [comment]
Updated:
- Added explicit option to chose provider
- fixed tests
- increased awesomeness by 3%
- removed unnecessary changes

File: src/internal/RemoteConfig.d [hunk 8]
```diff
@@ -0,0 +1,202 @@
+/*******************************************************************************
+
+    Helper methods to access the octod/github API
+
+    Copyright:
+        Copyright (c) 2017 sociomantic labs GmbH. All rights reserved.
```
Comment (leandro-lucarella-sociomantic) [inline src/internal/RemoteConfig.d:6]: Oh, we need to use the new dunnhumby copyright notice!

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
Except for the copyright notice (which can be addressed globally in a separate PR, or a separate commit), LGTM. If there are no more comments until Monday I would say we can merge and improve / add the missing stuff in future PRs.

**Comment by leandro-lucarella-sociomantic** [comment]
:tada: 

How was your experience with this, how hard to you think it will be to port the other tools too, like the auto-pr?

**Comment by mathias-baumann-sociomantic** [comment]
auto-pr will be very hard because it makes heavy use of the GraphQL API, something which isn't available in gitlab (yet!)

**Comment by leandro-lucarella-sociomantic** [comment]
Any ideas on how to handle this situation? Does this mean we just experienced the rise and death of auto-pr? @mihails-strasuns any suggestions?

**Comment by mihails-strasuns** [comment]
GitLab does work on graphql api too (alpha right now), so it may be feasible to wait until it exposes required information. It is also possible to resort to regular API as a fallback - will likely end up being 100x times slower but that is not a disaster for a tool that runs once a day.
